test/framework/src/main/java/org/elasticsearch/plugins/MockPluginsService.java

```diff
@@ -91,8 +92,21 @@ public <T> List<? extends T> loadServiceProviders(Class<T> service) {
         // all plugins in MockNode, because all plugins are loaded by the same class loader.
         Set<T> result = new HashSet<>();
 
+        Collection<Class<? extends T>> spiClasses = null;
```

Comment:
Since SPI is based on scanning the classpath (at least the META-INF based SPI which this method is based on), the resulting found SPI classes should be the same for the lifetime of the test JVM, they won't change across constructions MockPluginsService. Therefore, I think it would be better to do the SPI iteration once and stash it statically (say in a SetOnce). The extension creation can still happen on each call to this method, but the scanning is the expensive part, as I understand it.